providers/base/tests/test_gpio_control_test.py

```diff
@@ -0,0 +1,351 @@
+
+        gpio_no = "30"
+        gpio_controller = GPIOController("1", "1", "out", True)
+        gpio_controller._export(gpio_no)
+        mock_write.assert_called_once_with(
+            PosixPath('/sys/class/gpio/export'), gpio_no, False)
+
+    @patch("gpio_control_test.GPIOController.get_gpiochip_mapping")
+    @patch("gpio_control_test.GPIOController._write_node")
+    def test_unexport_gpio_node(self, mock_write, mock_mapping):
+        mock_mapping.return_value = {"1": "32"}
+
+        gpio_no = "30"
+        gpio_controller = GPIOController("1", "1", "out", True)
+        gpio_controller._unexport(gpio_no)
+        mock_write.assert_called_once_with(
+            PosixPath('/sys/class/gpio/unexport'), gpio_no, False)
+
+    @patch("gpio_control_test.GPIOController.get_gpiochip_mapping")
+    @patch("gpio_control_test.GPIOController._read_node")
+    @patch("gpio_control_test.GPIOController._node_exists")
+    def test_get_gpio_direction(self, mock_path, mock_path_get, mock_mapping):
+        mock_path.return_value = True
+        mock_path_get.return_value = "out"
+        mock_mapping.return_value = {"1": "32"}
+
+        gpio_controller = GPIOController("1", "1", "out", True)
+        gpio_controller.gpio_node = Path("fake-node")
+        self.assertEqual(gpio_controller.direction, "out")
+
+    @patch("gpio_control_test.GPIOController.get_gpiochip_mapping")
+    def test_set_gpio_direction_failed(self, mock_mapping):
+        mock_mapping.return_value = {"1": "32"}
+
+        with self.assertRaises(ValueError):
+            gpio_controller = GPIOController("1", "1", "out", True)
+            gpio_controller.direction = "wrong_direction"
+
+    @patch("gpio_control_test.GPIOController.get_gpiochip_mapping")
+    @patch("gpio_control_test.GPIOController._read_node")
+    @patch("gpio_control_test.GPIOController._node_exists")
+    def test_get_gpio_value(self, mock_path, mock_path_get, mock_mapping):
+        mock_path.return_value = True
+        mock_path_get.return_value = "1"
+        mock_mapping.return_value = {"1": "32"}
+
+        gpio_controller = GPIOController("1", "1", "out", True)
+        gpio_controller.gpio_node = Path("fake-node")
+        self.assertEqual(gpio_controller.value, "1")
+
+    @patch("gpio_control_test.GPIOController.get_gpiochip_mapping")
+    def test_set_gpio_value_failed(self, mock_mapping):
+        mock_mapping.return_value = {"1": "32"}
+
+        with self.assertRaises(ValueError):
+            gpio_controller = GPIOController("1", "1", "out", True)
+            gpio_controller.value = "wrong_value"
+
+    @patch("gpio_control_test.GPIOController.get_gpiochip_mapping")
+    @patch("gpio_control_test.GPIOController.off")
+    @patch("gpio_control_test.GPIOController.on")
+    def test_blinking_function(self, mock_on, mock_off, mock_mapping):
+        mock_mapping.return_value = {"1": "32"}
+
+        gpio_controller = GPIOController("1", "1", "out", True)
+        gpio_controller.blinking(1, 0.5)
```

Comment:
```suggestion
        gpio_controller.blinking(0.0001, 0.0001)
```

let's keep this as short as possible, otherwise this will slow down every unit test run in the future.